diesel_derives/src/insertable.rs

```diff
@@ -86,41 +70,85 @@ fn field_ty(
     field: &Field,
     table_name: &syn::Ident,
     lifetime: Option<proc_macro2::TokenStream>,
-) -> syn::Type {
-    if field.has_flag("embed") {
+) -> Result<syn::Type, Diagnostic> {
+    let ty = if field.has_flag("embed") {
         let field_ty = &field.ty;
         parse_quote!(#lifetime #field_ty)
     } else {
-        let inner_ty = inner_of_option_ty(&field.ty);
-        let column_name = field.column_name();
-        parse_quote!(
-            std::option::Option<diesel::dsl::Eq<
-                #table_name::#column_name,
-                #lifetime #inner_ty,
-            >>
-        )
-    }
+        match field.ty_for_serialize()? {
+            Some(ty) => {
+                let inner_ty = inner_of_option_ty(&ty);
+                let column_name = field.column_name();
+                parse_quote!(
+                std::option::Option<diesel::dsl::Eq<
+                    #table_name::#column_name,
+                    #inner_ty,
+                >>
+            )
+            }
+            None => {
+                let ty = &field.ty;
+
+                let inner_ty = inner_of_option_ty(&ty);
+                let column_name = field.column_name();
+                parse_quote!(
+                    std::option::Option<diesel::dsl::Eq<
+                        #table_name::#column_name,
+                        #lifetime #inner_ty,
+                    >>
+                )
+            }
+        }
+    };
+
+    Ok(ty)
 }
 
 fn field_expr(
     field: &Field,
     table_name: &syn::Ident,
     lifetime: Option<proc_macro2::TokenStream>,
-) -> syn::Expr {
-    let field_access = field.name.access();
-    if field.has_flag("embed") {
-        parse_quote!(#lifetime self#field_access)
-    } else {
-        let column_name = field.column_name();
-        let column: syn::Expr = parse_quote!(#table_name::#column_name);
-        if is_option_ty(&field.ty) {
-            if lifetime.is_some() {
-                parse_quote!(self#field_access.as_ref().map(|x| #column.eq(x)))
+) -> Result<syn::Expr, Diagnostic> {
+    let expr = match field.ty_for_serialize()? {
+        Some(ty) => {
+            let field_access = field.name.access();
+
+            if field.has_flag("embed") {
+                panic!("unsupported combination")
+            } else {
+                let column_name = field.column_name();
+                let column: syn::Expr = parse_quote!(#table_name::#column_name);
+
+                if is_option_ty(&ty) {
+                    if lifetime.is_some() {
+                        unimplemented!()
+                    } else {
+                        parse_quote!(self#field_access.map(|x| #column.eq(#ty::from(x))))
```

Comment:
We should not depend on the standard prelude here, but instead use UFCS otherwise this will break as soon as someone does something fancy…
```suggestion
                        parse_quote!(self#field_access.map(|x| #column.eq(<#ty as std::convert::From<#field_type>>::from(x))))
```
I think we should call `From<Type>` for owned types, and `From<&Type>` for types that are stored as reference in the original insertable struct.